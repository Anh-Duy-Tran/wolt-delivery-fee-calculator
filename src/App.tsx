import { SetStateAction, useState } from 'react';
import { FeeResult } from './components/FeeResult/FeeResult';
import { OrderDetailsForm } from './components/OrderDetailsForm';
import { FeeCalculatorReturnType } from './utils/deliveryFeeCaculator';
import { OrderDetailFormType } from './components/OrderDetailsForm/OrderDetailsForm';
import { Introduction } from './components/Introduction';

function App() {
  const [calculatedOrderDetails, setCalculatedOrderDetails] =
    useState<OrderDetailFormType>();
  const [feeResult, setFeeResult] = useState<FeeCalculatorReturnType>();

  const handleResult = (
    result: SetStateAction<FeeCalculatorReturnType | undefined>
  ) => setFeeResult(result);

  const saveCalculatedOrderDetails = (orderDetails: OrderDetailFormType) =>
    setCalculatedOrderDetails(orderDetails);

  return (
    <div id="background">
      <div className="flex h-full w-full flex-col items-center overflow-auto pt-[30px]">
        <h1>DELIVERY FEE CALCULATION</h1>
        <div className="m-8 grid w-[70%] max-w-[1450px] grid-cols-1 gap-4 tablet:grid-cols-2">
          <div>
            <OrderDetailsForm
              saveCalculatedOrderDetails={saveCalculatedOrderDetails}
              handleResult={handleResult}
            />
          </div>
          <div className="flex flex-col gap-4">
            <FeeResult
              orderDetails={calculatedOrderDetails}
              result={feeResult}
            />
            <Introduction />
          </div>
        </div>
      </div>
    </div>
  );
}

export default App;
