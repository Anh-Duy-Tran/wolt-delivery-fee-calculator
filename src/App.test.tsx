import { render, screen } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import App from './App';

describe('App Component', () => {
  it('render correctly', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should contains 3 card (form, result, introduction)', () => {
    render(<App />);
    expect(screen.getAllByTestId('card').length).toBe(3);
  });
});
