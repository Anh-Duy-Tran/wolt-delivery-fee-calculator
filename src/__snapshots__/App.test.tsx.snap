// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > render correctly 1`] = `
<DocumentFragment>
  <div
    id="background"
  >
    <div
      class="flex h-full w-full flex-col items-center overflow-auto pt-[30px]"
    >
      <h1>
        DELIVERY FEE CALCULATION
      </h1>
      <div
        class="m-8 grid w-[70%] max-w-[1450px] grid-cols-1 gap-4 tablet:grid-cols-2"
      >
        <div>
          <div
            class="flex flex-col gap-4 w-full rounded-xl bg-white p-5 shadow-xl"
            data-testid="card"
          >
            <h2>
              ORDER INFORMATION
            </h2>
            <form
              class="flex flex-col gap-2"
              novalidate=""
            >
              <div>
                <div
                  class="relative flex w-full items-center justify-center"
                >
                  <input
                    class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
                    data-testid="cartValue"
                    inputmode="numeric"
                    name="cartValue"
                    step="0.1"
                    type="text"
                  />
                  <label
                    class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] false"
                  >
                    Cart Value (EUR)
                  </label>
                </div>
              </div>
              <div>
                <div
                  class="relative flex w-full items-center justify-center"
                >
                  <input
                    class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
                    data-testid="deliveryDistance"
                    inputmode="numeric"
                    name="deliveryDistance"
                    step="10"
                    type="text"
                  />
                  <label
                    class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] false"
                  >
                    Delivery Distance (Meter)
                  </label>
                </div>
              </div>
              <div>
                <div
                  class="relative flex w-full items-center justify-center"
                >
                  <input
                    class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
                    data-testid="numberOfItems"
                    inputmode="numeric"
                    name="numberOfItems"
                    type="text"
                  />
                  <label
                    class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] false"
                  >
                    Number of Items
                  </label>
                </div>
              </div>
              <div>
                <div
                  class="relative flex w-full items-center justify-center"
                >
                  <input
                    class="peer w-full rounded-md border-2 border-wolt-grey p-4 pb-[6px] pt-[22px] outline-none transition placeholder:opacity-0 hover:border-wolt-blue focus:border-wolt-blue focus:shadow-border-like focus:placeholder:opacity-100"
                    data-testid="orderTime"
                    name="orderTime"
                    placeholder="DD.MM.YYYY"
                    type="datetime-local"
                  />
                  <label
                    class="linear duration-config pointer-events-none absolute w-full origin-left translate-x-4 scale-100 
        select-none opacity-60 transition-all peer-focus:-translate-y-3 peer-focus:scale-[65%] -translate-y-3 scale-[65%]"
                  >
                    Time
                  </label>
                </div>
              </div>
              <button
                class="mt-2 rounded-[6px] bg-wolt-blue p-3 text-white"
                data-testid="calculateFee"
                type="submit"
              >
                CALCULATE FEE
              </button>
            </form>
          </div>
        </div>
        <div
          class="flex flex-col gap-4"
        >
          <div
            class="flex flex-col gap-4 w-full rounded-xl bg-white p-5 shadow-xl"
            data-testid="card"
          >
            <h2>
              CALCULATION RESULT
            </h2>
            <div
              class="w-full text-center italic text-black/50"
            >
              Fill in order details and press calculate to get result.
            </div>
          </div>
          <div
            class="flex flex-col gap-4 w-full rounded-xl bg-white p-5 shadow-xl"
            data-testid="card"
          >
            <h2>
              INTRODUCTION
            </h2>
            <div
              class="flex items-center gap-2"
            >
              <div
                class="relative"
              >
                <div
                  class="peer rounded ring-wolt-blue focus:ring-2"
                  tabindex="0"
                >
                  <a
                    href="https://codecov.io/gh/Anh-Duy-Tran/wolt-delivery-fee-calculation"
                  >
                    <img
                      src="https://codecov.io/gh/Anh-Duy-Tran/wolt-delivery-fee-calculation/graph/badge.svg?token=CJQV0NZ99N"
                    />
                  </a>
                </div>
                <div
                  class="absolute right-0 z-50 mt-2 w-[350px] scale-75 rounded-md bg-slate-700 p-3 text-sm text-white opacity-0 transition peer-hover:scale-100 peer-hover:opacity-100 peer-focus:scale-100 peer-focus:opacity-100"
                  data-testid="tooltip"
                >
                  <p>
                    Codecov Report: Shows the percentage of code covered by tests.
                  </p>
                </div>
              </div>
              <div
                class="relative"
              >
                <div
                  class="peer rounded ring-wolt-blue focus:ring-2"
                  tabindex="0"
                >
                  <a
                    href="https://cloud.cypress.io/projects/jsr1iq/runs"
                  >
                    <img
                      src="https://img.shields.io/endpoint?url=https://cloud.cypress.io/badge/simple/jsr1iq/main&style=flat&logo=cypress"
                    />
                  </a>
                </div>
                <div
                  class="absolute right-0 z-50 mt-2 w-[350px] scale-75 rounded-md bg-slate-700 p-3 text-sm text-white opacity-0 transition peer-hover:scale-100 peer-hover:opacity-100 peer-focus:scale-100 peer-focus:opacity-100"
                  data-testid="tooltip"
                >
                  <p>
                    Cypress Status: Indicates the results of automated end-to-end tests using Cypress.
                  </p>
                </div>
              </div>
            </div>
            <div>
              <p>
                Integrated CI/CD pipelines with Codecov and Cypress badges display test results live, ensuring reliability.
              </p>
              <br />
              <img
                src="./intro-tooltip.gif"
              />
              <p
                class="text-center"
              >
                please try it out !
              </p>
              <br />
              <p>
                My project does more than just calculate fees - it demystifies them. Each fee component is clearly explained through interactive, reusable tooltips, offering users insight into their delivery costs with just a hover or click. I've also prioritized accessibility, making the app easily navigable and understandable for all users.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
